### PR TITLE
Implement Row status

### PR DESCRIPTION
**Pull request by derrell**
I'd welcome comments on the approach begun here. (It'll be easier to see the few changes here once the MAX-ACCESS PR is merged, as this builds on top of it.)

### CODE REVIEW

**Comment by markabrahams** [comment]
Hi @derrell - thanks for doing this!  FYI - I've just merged the MAX-ACCESS PR with a few modifications, so you might want to merge the new master back.

Haven't had a chance to look at this PR yet, but will do so in the next day or so.

**Comment by markabrahams** [comment]
Hi @derrell - are you OK to merge in the latest master changes (it's just the MAX-ACCESS piece really) here before I take a look?

**Comment by derrell** [comment]
Absolutely! Done. (I'm working on this right now.)

**Comment by derrell** [comment]
@markabrahams DEFVAL / read-create handling for scalars is working. Table row auto-insert and delete via RowStatus is mostly implemented but not yet tested. The general concept, in both cases, is that the library doesn't know enough to provide default values if DEFVAL isn't specified, so user-registered callbacks are used so that the application can fill in appropriate values. I'd appreciate comments! Thanks.

File: index.js [hunk 1]
```diff
@@ -3222,6 +3240,27 @@ ModuleStore.prototype.getProvidersForModule = function (moduleName) {
...
+
+								// Mark this column as being rowStatus
+								columnDefinition.rowStatus = true;
+
+								// Also keep track of the rowStatus
+								// column on a provider basis. (Note
+								// that this will not work correctly
+								// if there are multiple columns in a
+								// single table with syntax RowStatus.
+								// Only the last one will be honored.)
```
Comment (markabrahams) [inline index.js:3261]: That's fine - the intention is to have a single RowStatus column per table, so this is the only case we need to cater for..  From RFC 2578 Section 7.1.12.1:

"For newly-defined conceptual rows which allow the creation of new object instances and/or the deletion of existing object instances, there should be one columnar object with a SYNTAX clause value of RowStatus."

File: index.js [hunk 2]
```diff
@@ -4368,9 +4620,31 @@ Agent.prototype.request = function (requestMessage, rinfo) {
 					oid: requestPdu.varbinds[i].oid
 				});
 
-				if ( requestPdu.type == PduType.SetRequest ) {
-					mibRequests[i].setType = requestPdu.varbinds[i].type;
-					mibRequests[i].setValue = requestPdu.varbinds[i].value;
+				if ( requestPdu.type == PduType.SetRequest && instanceNode.address.length >= 2 ) {
+					var addrLen = instanceNode.address.length;
+					var row = instanceNode.address[addrLen - 1];
```
Comment (markabrahams) [inline index.js:4625]: It's important to note that the row identifier in an OID is not necessarily the single final number in the OID.  There are several scenarios (composite index, non-integer index, etc.) where this isn't the case - the row could be the last several numbers in the OID.  The column, however, is always a single integer from the OID.  For the column, there's already MibNode.getTableColumnFromInstanceNode() to get this.  A similar function could be written to get the row index (would be an array of integers).

File: index.js [hunk 3]
```diff
@@ -4368,9 +4620,31 @@ Agent.prototype.request = function (requestMessage, rinfo) {
...
+					if ( requestPdu.varbind.value === RowStatus["destroy"] &&
+						typeof provider.rowStatusColumn == "number" &&
+						column === provider.rowStatusColumn ) {
+
+						// Delete the table row
+						this.mib.deleteTableRow ( name, row );
+
+						//
+						// TODO: What goes in mibRequests[i]?????
+						//
```
Comment (markabrahams) [inline index.js:4643]: These are meant to reflect the incoming varbind type and value, so I don't think you need to deviate from the existing code there:
```
mibRequests[i].setType = requestPdu.varbinds[i].type;
mibRequests[i].setValue = requestPdu.varbinds[i].value;
```

File: index.js [hunk 4]
```diff
@@ -4259,6 +4319,189 @@ Agent.prototype.onMsg = function (buffer, rinfo) {
...
+	len = address.length;
+	if ( len < 3 ) {
+		console.log("OID is fubar; don't try further");
+		return undefined;
+	}
+
+	// Extract the row and column number before we start searching
+	// upwards for the provider
+	row = address[len - 1];
+	column = address[len - 2];
```
Comment (markabrahams) [inline index.js:4397]: See comment below on this, but in short, the column calculation needs to be MibNode.getTableColumnFromInstanceNode()

and the row needs to be an integer array, returned by a new but similar call like:
MibNode.getRowIndexFromInstanceNode()

**Comment by derrell** [comment]
Thanks for the comments! (And you're reviewing this at... what?... 2am your time?!?)

As you have probably guessed by now, I have lots of experience and knowledge, but somehow my path and SNMP's had never crossed (other than in casual conversation) until a few weeks ago, so I'm still coming up to speed. I need to read about indexes so I understand the row number issues. (I guess I'm getting sufficiently into this stuff that I should really take the time to read the whole spec...)

The problem I think I see with your suggestion, though, is that we don't yet have an instance node at the time that you're saying I should use  `MibNode.getTableColumnFromInstanceNode() ` and its new equivalent for rows. I think we'll need two new methods, that take an OID as an argument, rather than called on an instance node. I'll look into it, though. Maybe there is already code there to handle that.

So it sounds like you're aligned with my general methodology here, yes?

Thanks again!

**Comment by markabrahams** [comment]
Yeah it's a bit late here - and it shows! :-)  You're right - without an instance node, my suggestion is bogus!  And also right that I think a couple of Mib or MibNode methods to determine some of this.  Mib.getTreeNode() is almost what you want to get the provider node (I favour dynamically hitting the tree to determine the provider instead of caching), except it's geared to help getNext,  Maybe something like this to get the nearest ancestor (which should be the provider node for a new table create OID)

```
Mib.prototype.getNearestAncestor = function (address) {
	var i;
	var node;

	node = this.root;
	for (i = 0; i < address.length; i++) {
		if ( ! node.children.hasOwnProperty (address[i])) {
			return node;
		}
		node = node.children[address[i]];
	}

};  
```

Then it's easy to calculate the column (the next integer after the provider OID) and the row index array (everything left).

**Comment by markabrahams** [comment]
Your general methodology looks good to me so far.  I haven't done a very thorough job of understanding the RowStatus nuances from the RFC and matching them with your code yet, but will endeavour to do so maybe after some sleep!

**Comment by derrell** [comment]
Yeah, I'm far enough into this that I will re-read the RowStatus text to make sure this implementation is conformant. I typically do that at the end of a development as well, to ensure I didn't miss anything. In this case, much of the conformance is being left to the using application, by design, implemented in its callback. I just want to ensure that I provide the capability for the developer to implement in a conformant fashion (or shoot themself in the foot, should they so desire. :-) )

**Comment by derrell** [comment]
@markabrahams RowStatus to add and delete table rows is working, at least for the simple case of a table with a single column integer index. It has potential, I believe, to work with more sophisticated index structures, but I don't know how to test it. (I believe I've removed all assumptions, except in the example code, that the index is a single integer column.)

Run `node example/snmp-agent-rowstatus.js` and then
- snmpwalk -v2c -c public localhost:1611 1.0.15628.4.1
- snmpset -v2c -c public localhost:1611 rsuSetSlaveRowStatus.2 i 4
- snmpwalk -v2c -c public localhost:1611 1.0.15628.4.1
- snmpset -v2c -c public localhost:1611 rsuSetSlaveRowStatus.2 i 6
- snmpwalk -v2c -c public localhost:1611 1.0.15628.4.1

I need to do much more testing, and there are some comment-embedded concerns I have that I'd like your opinion on.

Watchathink?

**Comment by derrell** [comment]
@markabrahams I've gotten rid of the separate `example/snmp-agent-rowstatus.js` and integrated simple examples of RowStatus and scalar read-create into the main `snmp-agent.js` example.

**Comment by derrell** [comment]
@markabrahams I've removed the WIP indicator. This is looking pretty good and tests are showing good results. It's ready for your review.

**Comment by markabrahams** [comment]
Nice one @derrell!  I'll try to get a good look at this over the next day or so.

**Comment by markabrahams** [comment]
Hi @derrell - I've reviewed the PR and first of all, great work!  Looks really good and my initial testing checks out really well.

Sorry I didn't address your testing questions above before now, but I've tested against the example/snmp-agent.js like this:

```
mark@hope:~/node-net-snmp$ snmpset -v 2c -c private localhost:1161 1.3.6.1.2.1.2.2.1.99.4 i 5
IF-MIB::ifEntry.99.4 = INTEGER: 2
```
which worked as expected i.e. created a new row with default values for ifDescr and ifType, and correctly left ifStatus as noInService.

Further, this test with a supplied ifDescr worked brilliantly as well:
```
mark@hope:~/node-net-snmp$ snmpset -v 2c -c private localhost:1161 1.3.6.1.2.1.2.2.1.99.5 i 4 1.3.6.1.2.1.2.2.1.2.5 s "my new if"
IF-MIB::ifEntry.99.5 = INTEGER: 1
IF-MIB::ifDescr.5 = STRING: my new if
```
i.e. correctly created a new row with the default value for ifType, my supplied value for ifDescr, and correctly left ifStatus as active.

And deleting also worked:
```
mark@hope:~/node-net-snmp$ snmpset -v 2c -c private localhost:1161 1.3.6.1.2.1.2.2.1.99.3 i 6
IF-MIB::ifEntry.99.3 = INTEGER: 1
```

So I think the base functionality is there and working great!

A few further thoughts on the current state:

(1) I expressed a desire earlier to find an OID's provider dynamically without a cache.  However, having seen your Proxy/cache implementation, which is small and clean, I'm not unhappy to proceed with that approach.  For some reason, were I implementing, I'd still do the dynamic lookup approach.  I think I've been burned by too many cache issues in the past!

(2) I wonder about the create handlers.  At the moment, it seems to be too much boilerplate and also internal-working kinda code to push onto the library user.  The setScalarReadCreateHandler and setTableRowStatusHandler functions you've defined in example/snmp-agent.js seem like generically useful enough code (setting defaults, returning the appropriate RowStatus value) that this should be moved to behaviour actioned by the library itself.  My angle here is to make the create handlers optional, like the library does for current provider request handlers.  So if you don't supply a create handler, the library should do the "expected" thing by default e.g. set all default column values, set the RowStatus to active / notInService as appropriate.  Pretty much the code you have in the example/snmp-agent.js handlers.

(3) I'm also wondering if we go the way of (2), whether there needs to be create handlers at all.  Each create operation is going to have the set handler invoked anyway, and any further custom creation code could conceivable go into the existing set handler function, which has access to all the same information (the RowStatus operation (createAndGo or createAndWait) and row index, and the possible defaults from the provider are all you have at this point, as no other custom column values exist or have been supplied at this point).  The up-sides are (a) simplicity of sorts (but see below) - one less handler for users to understand - and (b) consistency in that there is a single handler mechanism for at the protocol action level, irrespective of the contents of the requests - whether they're RowStatus operations, etc.  The down-sides are that (c) there would be a possible race condition between a default-valued row on create, before potentially "fixing" the default values to custom values at a (very) slightly later point in the normal set handler, and (d) creating custom code that is changing default values I think is messier for a user than creating the row from scratch.  So pros and cons there.

(4) If we do stick with explicit, separate handlers, to align with the other provider request handlers, I'd prefer these to be defined on the provider themselves, as opposed to agent-wide.  This just aligns more with the current handlers, which are all specified on the provider. e.g. a provider.createHandler function.

(5) Once the final implementation is settled, we'll need to add any appropriate documentation on RowStatus to the README as well, but this can wait for now!

Ultimately, we'll need to settle on those ^^^ but hopefully we can do that relatively quickly and get this merged, as the functionality you've implemented really is quite cool! :-)  Let me know your thoughts.

**Comment by derrell** [comment]
Hi @markabrahams. I'm glad the basic functionality is working for you as it had been for me during testing.

Regarding the callback to the app to get values... My early thought was to keep everything internal to the library, for ease of use. What I realized, though, is that the library simply doesn't know enough about the application to be able to pick reasonable default values. I'll try to explain my rationale by way of example.

Consider a MIB with a table that controls robots. Adding a new row in the table is bringing a new robot online. The table might be long-ago defined and in use, like this:

```
robotControlTable OBJECT-TYPE
    SYNTAX SEQUENCE OF RobotControlEntry
    MAX-ACCESS  not-accessible
    STATUS      current
    DESCRIPTION
        "Controls a set of robots."
    ::= { robot 1 }

robotControlEntry OBJECT-TYPE
    SYNTAX RobotControlEntry
    MAX-ACCESS  not-accessible
    STATUS      current
    DESCRIPTION
        "A row controlling a specific robot ."
    INDEX   { robotIndex }
    ::= {robotControlTable 1 }

RobotControlEntry ::= SEQUENCE {
    robotIndex              INTEGER,
    robotSpeed             INTEGER,
    robotStatus             RowStatus
}

robotIndex OBJECT-TYPE
    SYNTAX       INTEGER
    MAX-ACCESS   not-accessible
    STATUS       current
    DESCRIPTION
        "Robot Index"
    ::= { robotControlEntry 1 }

robotSpeed OBJECT-TYPE
    SYNTAX       INTEGER { reverseFast(1), reverseSlow(2), stop(3), forwardSlow(4), forwardFast(5) }
    MAX-ACCESS   read-create
    STATUS       current
    DESCRIPTION
        "Start the robot moving at the specified direction and speed"
    ::= { robotControlEntry 2 }

robotStatus OBJECT-TYPE
    SYNTAX       RowStatus
    MAX-ACCESS   read-create
    STATUS       current
    DESCRIPTION
        "Create (4) or Destroy (6) row entry, etc."
    ::= { robotControlEntry 3 }
```
So now we consider what happens when a new RobotControlEntry is created. How does our library library select default values for each field? In a much more complicated implementation, the concept of RowStatus's `notReady` would be used: until each column has been set by the manager, the table row is in a `notReady` state, and the library then has to keep track of, or look at, each column when any column is set, to see if the status column needs to change from `notReady` to `active`. That's a lot of error-prone code that I thought would be best avoided if possible. So the next alternative is to come up with default values for each column. If a DEFVAL is provided for a column, that's easy. If not, as in this case, what do we do? A naive implementation, such as I've put into the example code, might say, "Ok, if it's an integer, just assign it 0... and if there is a value constraint on an integer, just select the lowest value." Seems reasonable at first, but consider what happens in the above example, with robotSpeed. Using the algorithm I just described, it would select a value of 1, the smallest value in the constraint set. This would cause the robot indicated by the row index to instantly enter `reverseFast` state, potentially doing serious damage to the robot or its surroundings. Only the app could know that robotSpeed should have been set to value 3, `stop`, initially.

Now, having said all of that, the decision to add a callback to the user actually came before I decided to implement DEFVAL along the way. Now that I've done so, I think the vast majority -- maybe all -- of the need for a callback to provide a table row's column values may be eliminated, if we provide an easy way of setting default values in the provider. We'll still need a likely-wrong algorithm for setting default values when the app has neglected to provide them, but it's then the app developer's fault.

I'll look into this latter option over the next day or two. I agree that if there's a way to let the app control these values without the need for an additional handler, at least in the normal case, that's a better solution. I'll see if it's still necessary to allow for a handler, or if we can do without it entirely, obviated by setting defVal in the provider, via some API.

**Comment by markabrahams** [comment]
Yes I can see how that evolved!  I think we're agreed there then - if scalars providers use provider.defVal and table columns use tableColumn.defVal - again, pretty much what what you look for in example/snmp-agent.js - that would be a good solution.

I'm just about to improve the defVal MIB parsing - it comes up as a one-element array at the moment, and this should be just the single default value.  Will push a commit to your row-status branch shortly.

**Comment by markabrahams** [comment]
That commit fixes DEFVAL reading in to scalar providers and table column definitions of table providers.

**Comment by derrell** [comment]
Oh, as I look to implement this, I recall another reason I pushed the decision making up to the app. Currently, the MIB parser parses integer enumerations, but not integer sub-types (RFC-2578, Section 11.1), so the library, in a generic fashion, is not able to select a legal default value when subtypes are listed in the MIB definition, if `defVal` isn't specified. Would you like to implement integer sub-types? Or we require the app to set `defVal`? Or we leave it to the app?

**Comment by derrell** [comment]
I'm still really, really uncomfortable putting guesses into the library, even if sub-types are implemented. I'll implement an internal handler that depends on `defVal`. If a scalar provider doesn't have a `defVal`, or if every column of a table provider doesn't have a `defVal`, the internal handler will return `undefined` meaning that it's not to be auto-created. The user has options: either define `devVal` for each; or provide their own handler to override the internal one, that uses a mechanism other than (only) `defVal` to provide the values.

**Comment by derrell** [comment]
@markabrahams All right! No user-provided auto-create handler is required,now, in the (hopefully) typical case. The option for specifying a handler remains, should it be necessary for any particular application where `defVal` isn't adequate.

Pleaes review the `tableRowStatusHandlerInternal` method carefully, particularly as it applies to my usage of the row index array values. It does the correct thing for a single column index, but is untested on composite indexes, and I have no idea what to even expect when the index is external, implied, etc.

Documentation is fully updated and should be ready to go.

**Comment by markabrahams** [comment]
Great - thanks @derrell !  I'll review this in the next day or so.  Agree with the mandatory defVals for automatic row creating - seems like a good direction.

**Comment by markabrahams** [comment]
@derrell - on your comment about the `tableRowStatusHandlerInternal` with other index types, this should work fine as long as the new row is added with `addTableRow()`, which is what you've done.  So we're getting down to really fine details now!  My feedback on the current state is:

(1) The example/snmp-agent.js has a `provider.rowStatusColumn = 99`, and also a `provider.tableColumns[x].rowStatus = true`.  Since these both specify the same thing, we should just use one - removes the inconsistent state possibility (one defined but not the other) and easy for the user to get right.  If I'm reading the code right, it looks like only `provider.rowStatusColumn` is used anyway.  Don't mean to be a pain, but I think the nicest place for a prescriptive column definition is the `provider.tableColumns[x].rowStatus = true` place.   What do you think about changing over to use that (and getting rid of `provider.rowStatusColumn`)?  It then means you can know everything about a column by looking at the provider's column definition.

(2) The `scalarReadCreateHandler()` and `tableRowStatusHandler()` are global to the agent.  I'd prefer these to be per-provider, to keep consistency with the existing protocol handlers, which are provider-specific.  A provider would optionally define these, and if not defined on the provider in question, the internal (default value) handlers would be invoked.  If we were to go this way, the names for these handlers on the provider I might be inclined to collapse into a single name: `provider.createHandler`.  You also define whether a provider is scalar or table, so this wouldn't add much to the meaning, and `createHandler` is a bit more specific to its purpose than for example `tableRowStatusHandler` (as it doesn't handle all RowStatus operations e.g. deletes)

(3) `mib.setScalarDefaultValue()` and `mib.setTableRowDefaultValues()` are never called now.  Their existence confused me for a bit before I realised this.  Can these be deleted?

(4) I don't recall what the empty, unused `mib.getValueFromOidAddress()` is for - let's delete it.

(5) On the bare function `getRowIndexFromOid()` - just a stylistic thing with an eye on a possibly future move to ES6/TypeScript classes - but I'd favour putting any function without a "this" in it as a "class" (non-prototype) function on the "class" it closely matches.  So instead of moving from `Mib.prototype.getRowIndexFromOid()` to just `getRowIndexFromOid()`, I'd move it to `Mib.getRowIndexFromOid()` to keep some nominal association with the Mib class (this is analogous to a "static" function in a proper ES6/Typescript class).

(6) Again, purely stylistic, but in your `tc.forEach` line in `tableRowStatusHandlerInternal()`, I prefer referencing the `columnInfo` object directly in this line over `tc[index]` i.e.
`} else if ( "defVal" in columnInfo ) {`

Apart from (1) and (2), which are more design-level, the rest are very minor, and more stylistic things than absolute necessities.  Let me know your thoughts on this.  Also, I'm able to implement this stuff, or equally happy for you to tackle it!

I'll just push a really small master merge into this branch as well (was mainly used for my reviewing purposes, but hey it needed to be done anyway!)

**Comment by derrell** [comment]
Thanks for the thorough review! Comments below...

> (1) The example/snmp-agent.js has a `provider.rowStatusColumn = 99`, and also a `provider.tableColumns[x].rowStatus = true`. Since these both specify the same thing, we should just use one - removes the inconsistent state possibility (one defined but not the other) and easy for the user to get right. If I'm reading the code right, it looks like only `provider.rowStatusColumn` is used anyway. Don't mean to be a pain, but I think the nicest place for a prescriptive column definition is the `provider.tableColumns[x].rowStatus = true` place. What do you think about changing over to use that (and getting rid of `provider.rowStatusColumn`)? It then means you can know everything about a column by looking at the provider's column definition.

Both are used. In some cases, we need to know at the provider level whether there is (and which is) a RowStatus column; and in other cases, we need that knowledge at the column level. That said, it's an optimization to have it in both places. My impression is that the providers will most typically be created from `getProvidersForModule` and not hand-crafted in typical real-world use (that's certainly the case in my own app), so the two would not get out of sync and it was an easy optimization. I have no problem calculating `rowStatusColumn` when it is needed, though, since these functions are not called at great frequency. The optimization is likely unnecessary.

> (2) The `scalarReadCreateHandler()` and `tableRowStatusHandler()` are global to the agent. I'd prefer these to be per-provider, to keep consistency with the existing protocol handlers, which are provider-specific. A provider would optionally define these, and if not defined on the provider in question, the internal (default value) handlers would be invoked. If we were to go this way, the names for these handlers on the provider I might be inclined to collapse into a single name: `provider.createHandler`. You also define whether a provider is scalar or table, so this wouldn't add much to the meaning, and `createHandler` is a bit more specific to its purpose than for example `tableRowStatusHandler` (as it doesn't handle all RowStatus operations e.g. deletes)

No complaints. Sounds fine.

> (3) `mib.setScalarDefaultValue()` and `mib.setTableRowDefaultValues()` are never called now. Their existence confused me for a bit before I realised this. Can these be deleted?

Nope. These are indeed used. They're not intended for use within the library, however; they're intended for the library user to call. I use them extensively in my app. Since this implementation requires default values in order to make use of auto-creation, and providers are generally created by `getProvidersFromModule` where the MIB file does not include all of the needed default values, these functions are used to add the missing information rather than making changes to an authorititative MIB file. These methods are documented in the README as being for this purpose.

> (4) I don't recall what the empty, unused `mib.getValueFromOidAddress()` is for - let's delete it.

Ok.

> (5) On the bare function `getRowIndexFromOid()` - just a stylistic thing with an eye on a possibly future move to ES6/TypeScript classes - but I'd favour putting any function without a "this" in it as a "class" (non-prototype) function on the "class" it closely matches. So instead of moving from `Mib.prototype.getRowIndexFromOid()` to just `getRowIndexFromOid()`, I'd move it to `Mib.getRowIndexFromOid()` to keep some nominal association with the Mib class (this is analogous to a "static" function in a proper ES6/Typescript class).

Sure.

> (6) Again, purely stylistic, but in your `tc.forEach` line in `tableRowStatusHandlerInternal()`, I prefer referencing the `columnInfo` object directly in this line over `tc[index]` i.e.
> `} else if ( "defVal" in columnInfo ) {`

Agree. I'm not sure why I did it the way I did in that one specific case but not elsewhere in that block of code.

> ... Also, I'm able to implement this stuff, or equally happy for you to tackle it!

I'll make these changes today, freeing you up to merge #152 and and implement the MIB parser changes to complete #158 immediately thereafter. :-)

Thanks again.

**Comment by derrell** [comment]
@markabrahams Done!

**Comment by markabrahams** [comment]
Great - it seems we're aligned on these then!  I'll look to merge this shortly.

On (1), the library intention is definitely to support both hand-crafted providers as well as MIB-derived.  If a user wants to expose some of their own data, or otherwise doesn't have an existing MIB for the data, it's unlikely they'll craft a compliant MIB for it, and so will need to hand-craft a provider without a MIB.

On (3), understood!  I figured we had covered hand-crafted defaults (just include them in the provider definition) and MIB-specified defaults (read in by `store.getProvidersForModule()`), but didn't appreciate your use-case of MIB-specified defaults that you might want to override/augment with your own defaults, which is totally valid!  I also see you can use "undefined" to skip a default for a particular column, which is great.  Presumably you'd do that for index columns where a default doesn't make sense.

**Comment by markabrahams** [comment]
As I've gone to merge, there's one final change I've made - sorry I didn't pick this up earlier and run past you!  I've aligned the arguments to the `createHandler` more with the `handler` i.e. 'createHandler' now takes a single `createRequest` argument that has all of the info on it.  Advantages are that the `createHandler` will look more familiar for those having used the `handler`, and also it's nicer to extend if required by adding more fields.  In fact, we could add the normal handler mibRequest fields for convenience (or add ones later) like mibRequest.oid, mibRequest.setValue, etc.  These might be useful info for the `createHandler` anyway.

Anyway, merge done now, and published in version 3.1.1 of the npm (after a small oops in 3.1.0!).

Check it out and close this if you're happy (or let me know it not!)

**Comment by derrell** [comment]
LGTM. Thanks!

**Pull request merged**